UniversalAlgebra/Complexity/FiniteCSP.lagda

```diff
@@ -0,0 +1,138 @@
+---
+layout: default
+title : Complexity.FiniteCSP module (The Agda Universal Algebra Library)
+date : 2021-07-26
+author: [agda-algebras development team][]
+---
+
+### Constraint Satisfaction Problems
+
+\begin{code}
+
+{-# OPTIONS --without-K --exact-split --safe #-}
+
+module Complexity.FiniteCSP  where
+
+open import Agda.Builtin.Equality using ( _≡_ ; refl )
+open import Data.Bool             using ( Bool ; true ; false ; _∧_)
+open import Data.Fin.Base         using ( Fin ; toℕ ; fromℕ ; raise)
+open import Data.Vec              using ( Vec ; [] ; tabulate ; lookup ; _∷_ ; map )
+open import Data.Vec.Relation.Unary.All using ( All )
+open import Data.Nat              using ( ℕ ; zero ; suc ; _+_ )
+
+open import Agda.Primitive   using ( _⊔_ ; lsuc ; Level) renaming ( Set to Type )
+open import Function.Base    using ( _∘_ )
+open import Relation.Binary  using ( DecSetoid)
+
+private variable
+ α ℓ : Level
+
+\end{code}
+
+#### Constraints
+
+We start with a vector (of length n, say) of variables.  For simplicity, we'll use the natural
+numbers for variable symbols, so
+
+V : Vec ℕ n
+V = [0 1 2 ... n-1]
+
+We can use the following range function to construct the vector of variables.
+
+\begin{code}
+
+range : (n : ℕ) → Vec ℕ n
+range n = tabulate toℕ
+-- `range n` is 0 ∷ 1 ∷ 2 ∷ … ∷ n-1 ∷ []
+
+\end{code}
+
+Let `nvar` denote the number of variables.
+
+A *constraint* consists of
+
+1. a natural number `∣s∣` denoting the number of variables in the "scope" of the constraint;
+
+2. a scope vector,  `s : Vec (Fin nvar) ∣s∣` , where `s i` is `k` if `k` is the i-th variable
+   in the scope of the constraint.
+
+3. a constraint relation, rel, which is a collection of functions mapping (indices of)
+   variables in the scope to elements of the domain.
+
+To summarize, a constraint is a triple (∣s∣ , s , rel), where
+* ∣s∣ is the number of variables in scope s.
+* s is the scope function: s i ≡ v iff v is the i-th variable in scope s.
+* rel is the contraint relation: a collection of functions mapping indices
+  of scope variables to elements in the domain.
+
+\begin{code}
+
+open DecSetoid
+open Fin renaming (zero to fzer ; suc to fsuc)
+
+record Constraint (nvar : ℕ) (dom : Vec (DecSetoid α ℓ) nvar) : Type α where
+ field
+  ∣s∣ : Fin nvar                     -- The "number" of variables involved in the constraint.
+  s : Vec (Fin nvar) (toℕ ∣s∣)        -- Vec of variables involved in the constraint.
```

Comment:
So this encoding is going to be hard to work with. For one, it could repeat variables (as there's nothing preventing that).

Why not `Vec Bool nvar` ? Then `|s|` would be derivable by counting.  Then you can use the `T` type from stdlib to good effect.

Reply:
I don't really see what's wrong with repeating variables.

I thought about using `Vec Bool nvar`.  In fact, that's what I started with.  I don't see immediately why that would be better, but I'm willing to give it another try.

I'm not familiar with the `T` type, but I'll check it out.

Reply:
Oh, did you just mean `T` as in "top?"  Hahaha... I thought there was some special generic `T` type I was supposed to know about!

Reply:
No, I mean `T` as in the very last definition in https://agda.github.io/agda-stdlib/Data.Bool.Base.html

Reply:
Ah, now I see, you probably meant `T` as in the type at the bottom of `Data.Bool.Base` that maps `true` to top and `false` to bottom.  Okay, I try to make use of that and see how it helps.  Thanks!

Reply:
Okay, thanks!

Reply:
Repeated variables would mean that `|s|` isn't really the number of variables, just some upper bound. That will make writing certain things awkward.

Probably even better than `Vec Bool` would be some of Conor McBride's latest tricks for dealing with well-scoped terms. But that's seriously bleeding-edge stuff, so that's why I didn't suggest it.

Reply:
I see, sounds reasonable (re. repeated variables).

I'll check out Conor's stuff.  Thanks for the suggestion!

Reply:
I meant this paper in particular https://core.ac.uk/download/pdf/195292962.pdf .

Reply:
Thanks, I'll check it out.